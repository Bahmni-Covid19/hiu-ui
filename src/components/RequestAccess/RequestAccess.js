import React, { useState, useEffect } from "react";
import { Button, CircularProgress } from "@material-ui/core";
import RequestAccessStyles from "./RequestAccess.style";
import SearchPatient from "../SearchPatient";
import DateTimePicker from "./../DateTimePicker/DateTimePicker";
import DatePicker from "./../DateTimePicker/DatePicker";
import Grid from "@material-ui/core/Grid";
import InputLabel from "@material-ui/core/InputLabel";
import RequestType from "./../RequestType/RequestType";
import SimpleMenu from "./../SimpleMenu/SimpleMenu";
import getNextDay from "./../../utils/getNextDay";
import purposeTypes from "./../../constants/puposeTypes";
import requestTypes from "./../../constants/requestTypes";

const RequestAccess = ({
  onCreateConsent,
  patientId,
  loading,
  success,
  error
}) => {
  const [selectedPurposeValue, setSelectedPurposeValue] = React.useState(
    purposeTypes[0].value
  );
  const [selectedStartDate, setSelectedStartDate] = React.useState(new Date());
  const [selectedEndDate, setSelectedEndDate] = React.useState(new Date());
  const [selectedRequestTypes, setselectedRequestTypes] = React.useState({
    PatientHistory: false,
    Medications: false,
    DiagnosisLab: false,
    RadiologyLab: false,
    Condition: false
  });

  const [selectedExpiryDate, setSelectedExpiryDate] = React.useState(
    getNextDay()
  );

  const handlePITypeChange = name => event => {
    setselectedRequestTypes({
      ...selectedRequestTypes,
      [name]: event.target.checked
    });
  };
  const handleStarteDateChange = date => {
    var startDate = date;
    setSelectedStartDate(startDate);
  };
  const handleEndDateChange = date => {
    var startDate = selectedStartDate;
    var endDate = date;
    if (startDate.getTime() < endDate.getTime()) {
      setSelectedEndDate(endDate);
    }
  };
  const handleExpiryDateChange = date => {
    setSelectedExpiryDate(date);
  };

  const handlePurposeSelection = event => {
    setSelectedPurposeValue(event.target.value);
  };

  return (
    <RequestAccessStyles>
      <h2>Consent request form</h2>
      <form>
        <span className="label">All the fields are mandatory.</span>
        {error && (
          <span className="error">Error occured while creating consent.</span>
        )}
        {success && (
          <span className="success">Consent created successfylly.</span>
        )}
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label">
              Patient Identifier
            </InputLabel>
          </Grid>
          <Grid item xs={5}>
            <SearchPatient />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label">
              Purpose of request
            </InputLabel>
          </Grid>
          <Grid item xs={2}>
            <SimpleMenu
              menuItems={purposeTypes}
              handleChange={handlePurposeSelection}
              selectedValue={selectedPurposeValue}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label"> Request from </InputLabel>
          </Grid>
          <Grid item xs={2}>
            <DatePicker
              disableFuture={true}
              handleDateChange={handleStarteDateChange}
              selectedDate={selectedStartDate}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label"> Request to </InputLabel>
          </Grid>
          <Grid item xs={2}>
            <DatePicker
              disableFuture={true}
              minDate={selectedStartDate}
              handleDateChange={handleEndDateChange}
              selectedDate={selectedEndDate}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label"> Request type </InputLabel>
          </Grid>
          <Grid item xs={10}>
            <RequestType
              requestTypes={requestTypes}
              handleChange={handlePITypeChange}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label">Consent Expiry</InputLabel>
          </Grid>
          <Grid item xs={3}>
            <DateTimePicker
              handleDateChange={handleExpiryDateChange}
              selectedDate={selectedExpiryDate}
              disablePast="true"
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Button
            className="create-consent-button"
            variant="contained"
            color="primary"
            onClick={() => {
              onCreateConsent({
                patientId,
                selectedPurposeValue,
                selectedStartDate,
                selectedEndDate,
                selectedExpiryDate,
                selectedRequestTypes
              });
            }}
          >
            Request Consent
          </Button>
        </Grid>
      </form>
      {loading && (
        <div className="loader-container">
          <CircularProgress
            id="loader"
            className="loader"
            variant="indeterminate"
            disableShrink
            thickness={4}
          />
        </div>
      )}
    </RequestAccessStyles>
  );
};

SearchPatient.defaultProps = {
  success: false,
  error: false
};

export default RequestAccess;
