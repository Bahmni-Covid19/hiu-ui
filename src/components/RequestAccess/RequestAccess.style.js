import styled from "styled-components";

const RequestAccessStyles = styled.div`
  h2 {
    color: #000000;
    font-size: 24px;
    line-height: 24px;
    font-weight: 300;
    padding: 20px;
    margin-bottom: 30px;
    border-bottom: 1px solid #000000;
    font-family: "Roboto", sans-serif;
  }
  .label {
    display: block;
    color: #000;
    font-size: 16px;
    font-family: "Roboto", sans-serif;
  }
  .MuiGrid-container {
    padding: 10px 0;
  }
  .search-bar {
    vertical-align: center;
  }
  .auto-complete {
    color: #009688;
    .text-field {
      color: #009688;
    }
  }
  .error {
    display: block;
    padding: 10px 0;
    color: #f44336;
    font-size: 16px;
    font-family: "Roboto", sans-serif;
  }
`;

export default RequestAccessStyles;
