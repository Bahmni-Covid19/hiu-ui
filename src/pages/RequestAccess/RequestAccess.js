import React, { useState } from 'react';
import { Button, CircularProgress } from '@material-ui/core';
import Grid from '@material-ui/core/Grid';
import InputLabel from '@material-ui/core/InputLabel';
import RequestAccessStyles from './RequestAccess.style';
import SearchPatient from '../../components/SearchPatient';
import DateTimePicker from '../../components/DateTimePicker/DateTimePicker';
import DatePicker from '../../components/DateTimePicker/DatePicker';
import RequestType from '../../components/RequestType/RequestType';
import SimpleMenu from '../../components/SimpleMenu/SimpleMenu';
import getNextDay from '../../utils/getNextDay';

const RequestAccess = ({
  onCreateConsent, patientId, loading, error, purposesOfUse, hiTypes,
}) => {
  const [selectedStartDate, setSelectedStartDate] = useState(new Date());
  const [selectedEndDate, setSelectedEndDate] = useState(new Date());
  const usagePurposes = purposesOfUse.map((p) => ({ label: p.display, value: p.code }));
  const [selectedPurposeValue, setSelectedPurposeValue] = useState(
    usagePurposes.length > 0 ? usagePurposes[0].value : '',
  );
  const requestHiTypes = hiTypes.map((p) => ({ label: p.display, value: p.code }));
  const hiTypesInitialStates = Object.assign(...requestHiTypes.map((hiType) => ({ [hiType.value]: false })));
  const [selectedRequestTypes, setSelectedRequestTypes] = useState(hiTypesInitialStates);

  const [emptyPatientIDError, setEmptyPatientIDError] = useState(false);
  const [selectedExpiryDate, setSelectedExpiryDate] = useState(getNextDay());

  const isButtonEnabled = Object.values(selectedRequestTypes).some((x) => x);

  const handleHITypeChange = (name) => (event) => {
    setSelectedRequestTypes({
      ...selectedRequestTypes,
      [name]: event.target.checked,
    });
  };
  const handleStarteDateChange = (date) => {
    const startDate = date;
    setSelectedStartDate(startDate);
  };
  const handleEndDateChange = (date) => {
    const startDate = selectedStartDate;
    const endDate = date;
    if (startDate.getTime() < endDate.getTime()) {
      setSelectedEndDate(endDate);
    }
  };
  const handleExpiryDateChange = (date) => {
    setSelectedExpiryDate(date);
  };

  const handlePurposeSelection = (event) => {
    setSelectedPurposeValue(event.target.value);
  };

  return (
    <RequestAccessStyles>
      <h2>Consent request form</h2>
      <form>
        <span className="label">All the fields are mandatory.</span>
        {error && (
          <span className="error">Error occured while creating consent.</span>
        )}
        {emptyPatientIDError && (
          <span className="error">Please enter a patient identifier</span>
        )}
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label">
              Patient Identifier
            </InputLabel>
          </Grid>
          <Grid item xs={5}>
            <SearchPatient />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label">
              Purpose of request
            </InputLabel>
          </Grid>
          <Grid item xs={2}>
            <SimpleMenu
              menuItems={usagePurposes}
              handleChange={handlePurposeSelection}
              selectedValue={selectedPurposeValue}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label">
              Health info from
            </InputLabel>
          </Grid>
          <Grid item xs={2}>
            <DatePicker
              minDate="01/01/1900"
              disableFuture
              handleDateChange={handleStarteDateChange}
              selectedDate={selectedStartDate}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label">Health info to</InputLabel>
          </Grid>
          <Grid item xs={2}>
            <DatePicker
              disableFuture
              minDate={selectedStartDate}
              handleDateChange={handleEndDateChange}
              selectedDate={selectedEndDate}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label"> Request type </InputLabel>
          </Grid>
          <Grid item xs={10}>
            <RequestType
              requestTypes={requestHiTypes}
              handleChange={handleHITypeChange}
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Grid item xs={2}>
            <InputLabel className="text-field-label">Consent Expiry</InputLabel>
          </Grid>
          <Grid item xs={3}>
            <DateTimePicker
              handleDateChange={handleExpiryDateChange}
              selectedDate={selectedExpiryDate}
              disablePast="true"
            />
          </Grid>
        </Grid>
        <Grid container spacing={3} alignItems="center">
          <Button
            disabled={!isButtonEnabled}
            className="create-consent-button"
            variant="contained"
            color="primary"
            onClick={() => {
              if (!patientId) {
                setEmptyPatientIDError(true);
                return;
              }
              onCreateConsent({
                patientId,
                selectedPurposeValue,
                selectedStartDate,
                selectedEndDate,
                selectedExpiryDate,
                selectedRequestTypes,
              });
            }}
          >
            Request Consent
          </Button>
        </Grid>
      </form>
      {loading && (
        <div className="loader-container">
          <CircularProgress
            id="loader"
            className="loader"
            variant="indeterminate"
            disableShrink
            thickness={4}
          />
        </div>
      )}
    </RequestAccessStyles>
  );
};

SearchPatient.defaultProps = {
  success: false,
  error: false,
};

export default RequestAccess;
