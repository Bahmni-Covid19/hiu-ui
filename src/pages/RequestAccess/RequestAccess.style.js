import styled from "styled-components";

const RequestAccessStyles = styled.div`
  h2 {
    color: #000000;
    font-size: 24px;
    line-height: 24px;
    font-weight: 300;
    padding: 20px 0;
    margin-bottom: 25px;
    border-bottom: 1px solid #000000;
    font-family: "Roboto", sans-serif;
  }
  .text-field-label {
    color: #000;
  }
  .label {
    display: block;
    color: #000;
    font-size: 16px;
    font-weight: 300;
    font-family: "Roboto", sans-serif;
  }
  .MuiGrid-container {
    padding: 10px 0;
  }
  .search-bar {
    vertical-align: center;
  }
  .auto-complete {
    color: #009688;
    .text-field {
      color: #009688;
    }
  }
  .create-consent-button {
    margin-top: 25px;
  }
  .error {
    display: block;
    padding: 20px 0 0px;
    color: #f44336;
    font-size: 16px;
    font-weight: 300;
    font-family: "Roboto", sans-serif;
  }
  .loader-container {
    position: fixed;
    z-index: 99;
    left: 0;
    top: 75px;
    background: #fff;
    width: 100%;
    height: 100%;
    align-items: center;
    display: flex;
    justify-content: center;
  }
`;

export default RequestAccessStyles;
