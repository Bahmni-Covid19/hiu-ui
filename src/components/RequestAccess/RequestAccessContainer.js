import React from "react";
import { connect } from "react-redux";
import RequestAccess from "./RequestAccess";
import { onCreateConsent } from "../../redux/actions/createConsentActions";

const mapStateToProps = state => ({
  patientId: state.patientDetail.patientData.id,
  error: state.createConsent.error,
  success: state.createConsent.success
});

const mapDispatchToProps = {
  onCreateConsent
};

export default connect(mapStateToProps, mapDispatchToProps)(RequestAccess);
