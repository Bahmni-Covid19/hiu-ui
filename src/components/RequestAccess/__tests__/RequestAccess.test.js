import { shallow } from "enzyme";
import React from "react";
import RequestAccess from "../RequestAccess";

describe("Request Access", () => {
  it("should render properly ", () => {
    const wrapper = shallow(<RequestAccess />);
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
