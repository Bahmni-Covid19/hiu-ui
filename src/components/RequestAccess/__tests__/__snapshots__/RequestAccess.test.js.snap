// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Access should render properly  1`] = `
"<styled.div>
  <h2>
    Consent request form
  </h2>
  <form>
    <WithStyles(ForwardRef(Grid)) container={true} spacing={3} alignItems=\\"center\\">
      <WithStyles(ForwardRef(Grid)) item={true} xs={2}>
        <WithStyles(ForwardRef(InputLabel)) className=\\"text-field-label\\">
          Patient Identifier
        </WithStyles(ForwardRef(InputLabel))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid)) item={true} xs={5}>
        <Connect(SearchPatient) />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid)) container={true} spacing={3} alignItems=\\"center\\">
      <WithStyles(ForwardRef(Grid)) item={true} xs={2}>
        <WithStyles(ForwardRef(InputLabel)) className=\\"text-field-label\\">
          Purpose of request
        </WithStyles(ForwardRef(InputLabel))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid)) item={true} xs={2}>
        <SimpleMenu menuItems={{...}} handleChange={[Function: handlePurposeSelection]} selectedValue=\\"GeneralConsulting\\" />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid)) container={true} spacing={3} alignItems=\\"center\\">
      <WithStyles(ForwardRef(Grid)) item={true} xs={2}>
        <WithStyles(ForwardRef(InputLabel)) className=\\"text-field-label\\">
           Request from 
        </WithStyles(ForwardRef(InputLabel))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid)) item={true} xs={2}>
        <DatePicker disableFuture={true} handleDateChange={[Function: handleStarteDateChange]} selectedDate={[undefined]} />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid)) container={true} spacing={3} alignItems=\\"center\\">
      <WithStyles(ForwardRef(Grid)) item={true} xs={2}>
        <WithStyles(ForwardRef(InputLabel)) className=\\"text-field-label\\">
           Request to 
        </WithStyles(ForwardRef(InputLabel))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid)) item={true} xs={2}>
        <DatePicker disableFuture={true} handleDateChange={[Function: handleEndDateChange]} selectedDate={[undefined]} />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid)) container={true} spacing={3} alignItems=\\"center\\">
      <WithStyles(ForwardRef(Grid)) item={true} xs={2}>
        <WithStyles(ForwardRef(InputLabel)) className=\\"text-field-label\\">
           Request type 
        </WithStyles(ForwardRef(InputLabel))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid)) item={true} xs={10}>
        <RequestType requestTypes={{...}} handleChange={[Function: handlePITypeChange]} />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid)) container={true} spacing={3} alignItems=\\"center\\">
      <WithStyles(ForwardRef(Grid)) item={true} xs={2}>
        <WithStyles(ForwardRef(InputLabel)) className=\\"text-field-label\\">
          Consent Expiry
        </WithStyles(ForwardRef(InputLabel))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid)) item={true} xs={3}>
        <DateTimePicker handleDateChange={[Function: handleExpiryDateChange]} selectedDate={[undefined]} disableFuture={false} />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid)) container={true} spacing={3} alignItems=\\"center\\">
      <WithStyles(ForwardRef(Button)) variant=\\"contained\\" color=\\"primary\\" onClick={[Function: onClick]}>
        Request Consent
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </form>
</styled.div>"
`;
